### PR TITLE
Fix assets file naming to work across both *nix and windows

### LINKED ISSUE #213: add html anchors to each stdout line
Sometimes the captured stdout/stderr can be quite long and it would be very useful if users could share bookmarks that send user to the exact captured line.

Example: http://logs.openstack.org/73/645273/26/check/tox-molecule/87eb604/tox/reports.html

To enable this we may need to add line numbers to the output but this can be done in pure CSS.

### LINKED ISSUE #214: can not embed image on windows hosts using extra.png(image)
On a windows host, we cannot add an image to a report using `pytest_html.extras.png()`.
It returns the following error:
```
Testing started at 00:17 ...
C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\Scripts\python.exe "D:\Program Files\JetBrains\PyCharm Community Edition 2017.3.3\helpers\pycharm\_jb_pytest_runner.py" -- --html=report.html
Launching py.test with arguments --html=report.html in C:\Users\proto\PycharmProjects\pytest_html_extra_image

============================= test session starts =============================
platform win32 -- Python 3.6.4, pytest-5.0.0, py-1.8.0, pluggy-0.12.0
rootdir: C:\Users\proto\PycharmProjects\pytest_html_extra_image
plugins: html-1.21.1, metadata-1.8.0collected 1 item

tests\sample_test.py .
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\_pytest\main.py", line 213, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\_pytest\main.py", line 257, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\_pytest\main.py", line 278, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\_pytest\runner.py", line 72, in pytest_runtest_protocol
INTERNALERROR>     runtestprotocol(item, nextitem=nextitem)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\_pytest\runner.py", line 87, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\_pytest\runner.py", line 171, in call_and_report
INTERNALERROR>     hook.pytest_runtest_logreport(report=report)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\manager.py", line 87, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\manager.py", line 81, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "C:\Users\proto\PycharmProjects\pytest_html_extra_image\venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "c:\users\proto\pycharmprojects\pytest_html_extra_image\venv\lib\site-packages\pytest_html\plugin.py", line 509, in pytest_runtest_logreport
INTERNALERROR>     self.append_passed(report)
INTERNALERROR>   File "c:\users\proto\pycharmprojects\pytest_html_extra_image\venv\lib\site-packages\pytest_html\plugin.py", line 291, in append_passed
INTERNALERROR>     self._appendrow('Passed', report)
INTERNALERROR>   File "c:\users\proto\pycharmprojects\pytest_html_extra_image\venv\lib\site-packages\pytest_html\plugin.py", line 273, in _appendrow
INTERNALERROR>     result = self.TestResult(outcome, report, self.logfile, self.config)
INTERNALERROR>   File "c:\users\proto\pycharmprojects\pytest_html_extra_image\venv\lib\site-packages\pytest_html\plugin.py", line 119, in __init__
INTERNALERROR>     self.append_extra_html(extra, extra_index, test_index)
INTERNALERROR>   File "c:\users\proto\pycharmprojects\pytest_html_extra_image\venv\lib\site-packages\pytest_html\plugin.py", line 202, in append_extra_html
INTERNALERROR>     extra.get('extension'), 'wb')
INTERNALERROR>   File "c:\users\proto\pycharmprojects\pytest_html_extra_image\venv\lib\site-packages\pytest_html\plugin.py", line 169, in create_asset
INTERNALERROR>     with open(asset_path, mode, **kwargs) as f:
INTERNALERROR> OSError: [Errno 22] Invalid argument: 'C:\\Users\\proto\\PycharmProjects\\pytest_html_extra_image\\assets\\tests/sample_test.py::test00_bcfe5b9ea2948de983c2f4df731cb2e2.png'

========================== 1 passed in 12.62 seconds ==========================
Process finished with exit code 0
```
The behaviour is not observed on a Linux system
It might be due to:
- the double colons in the filename (forbidden on windows)
- the slash used to represent the test hierarchy

Here is a sample to reproduce the behaviour:
tests/conftest.py:
```python
import pytest

@pytest.mark.hookwrapper
def pytest_runtest_makereport(item, call):
    pytest_html = item.config.pluginmanager.getplugin('html')
    outcome = yield
    report = outcome.get_result()
    extra = getattr(report, 'extra', [])

    extra.append(pytest_html.extras.png(item.module.screenshot))

    report.extra = extra
```

tests/sample_test.py
```python
from selenium import webdriver


screenshot = None


def test():
    global screenshot
    driver = webdriver.Firefox()
    driver.get("https://docs.pytest.org/")

    screenshot = driver.find_element_by_tag_name("html").screenshot_as_base64

    driver.close()
```
then run:
```
pytest --html=report.html
```

### PR DESCRIPTION
**Pull request by BeyondEvil**
Instead of a combination of human readable filename and hash, let's try a filename with cross-OS friendly formatting.

Fixes #214
Fixes #213

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -32,3 +32,6 @@ exclude = .eggs,.tox
 
 [pytest]
 testpaths = testing
+markers =
+    hookwrapper
+    optionalhook
```
Comment (BeyondEvil) [inline tox.ini]: @davehunt I can't seem to find these being used (read) anywhere? 🤔

File: pytest_html/plugin.py [hunk 2]
```diff
@@ -167,21 +166,17 @@ def __lt__(self, other):
...
-            # 255 is the common max filename length on various filesystems,
-            # we subtract hash length, file extension length and 2 more
-            # characters for the underscore and dot
-            max_length = 255 - len(hex_digest) - len(file_extension) - 2
-            asset_file_name = "{0}_{1}.{2}".format(
-                hash_key[:max_length], hex_digest, file_extension
-            )
+            # 255 is the common max filename length on various filesystems
+            asset_file_name = "{}_{}_{}.{}".format(
+                self.test_id.replace(":", "_"),
```
Comment (davehunt) [inline pytest_html/plugin.py]: Are you confident that `test_id` can't contain other unsafe characters?

File: tox.ini (hunk 1 above)
Comment (davehunt) [reply tox.ini]: I don't recall why these were needed, perhaps they aren't necessary anymore. If the tests are passing, then I would assume it's safe to remove these.

Comment (BeyondEvil) [reply tox.ini]: Ok, I'll remove them in a separate PR for traceability then. 👍

File: pytest_html/plugin.py (hunk 2 above)
Comment (BeyondEvil) [reply pytest_html/plugin.py]: Now I am. (see the revised code) 😊 @davehunt 

Now **everything** that isn't `a-zA-Z0-9` or `_` or `.` is replaced with `_`

**Comment by BeyondEvil** [comment]
I've asked the people that filed the issues to do some testing on Windows since I don't have access to any windows machines (tried setting something up in a VM, but I've been away from Windows too long).

But it would be great if either of you have the possibility to do some smoke-testing? @RibeiroAna @davehunt

**Comment by EricDelaney** [comment]
@BeyondEvil Your change worked for me as well as it now saves the results correctly without error-ing out. Awesome

**Comment by BeyondEvil** [comment]
> @BeyondEvil Your change worked for me as well as it now saves the results correctly without error-ing out. Awesome

Thanks @EricDelaney !

Awesome, as soon as I can get someone to approve the PR, I'll create a new release. 👍

**Comment by BeyondEvil** [comment]
ping @davehunt 🙏

**Pull request merged**